### ISSUE TITLE
[RE] Effective Program Debloating via Reinforcement Learning

### ISSUE THREAD

**Issue by Kahss** [ReScience/submissions#48]
**Original article:** Effective Program Debloating via Reinforcement Learning

**PDF URL:** https://gitlab.com/onepoint/recherche-et-developpement/optimus/chisel-replication/-/blob/master/Replication_CHISEL.pdf
**Metadata URL:** https://gitlab.com/onepoint/recherche-et-developpement/optimus/chisel-replication/-/blob/master/metadata.yaml
**Code URL:** https://gitlab.com/onepoint/recherche-et-developpement/optimus/chisel-replication/-/tree/master

**Scientific domain:** Program Debloating and Machine Learning 
**Programming language:** Python
**Suggested editor:** Visual Studio Code or VIM

**Comment by rougier** [comment]
Thanks for your submission, we'll assign an editor soon.

**Comment by rougier** [comment]
@otizonaizit @benoit-girard @oliviaguest @gdetor Can any of you edit this submission?

**Comment by oliviaguest** [comment]
Happy to take this on. Can somebody explain how we manage GitLab submissions?

**Comment by rougier** [comment]
Thanks. And for gitlab, nothing special I think. If accepted, we'll ask archiving on software heritage that can deal with gitlab.

**Comment by khinsen** [comment]
Nothing special indeed. See [here](https://github.com/ReScience/submissions/issues/22) for a GitLab-based submission that we already handled.

**Comment by oliviaguest** [comment]
@Kahsius do you have any ideas for potential reviewers? Feel free to drop their names, or usernames (but without the @, to avoid needless notifications for them) here — thanks! 😊

**Comment by Kahss** [comment]
Hello everyone, and thank you @oliviaguest for considering my submission :smile: !
Actually I don't have any ideas for potential reviewers as it is my first submission to ReScience. Do I have to give some names for my paper to be reviewed ?

**Comment by oliviaguest** [comment]
@D-Maurel no, you do not *have* to — but it would be very useful.

**Comment by rougier** [comment]
@D-Maurel you can have a look at the board to suggest reviewers: http://rescience.github.io/board/
@oliviaguest  You can use notification to ask all reviewers at once

**Comment by oliviaguest** [comment]
@rougier I have already started asking people from my own contacts, for now.

**Comment by rougier** [comment]
I meant via ```@ ReScience / reviewers```

**Comment by Kahss** [comment]
Thank you @rougier for the link. After checking the reviewers list, I would suggest people with experience in Reinforcement Learning, namely:
- Konstantinos Chatzilygeroudis
- Emmanuel Bengio
- Amy Zhang
- Xuedong Shang

I hope it might help.

**Comment by oliviaguest** [comment]
👋  @emmanuelle @costashatz would either of you two be willing to review this? Thanks! 😊

**Comment by costashatz** [comment]
@oliviaguest thank you for the invitation. Unfortunately, this is a very busy period and I will not have time to review this submission correctly. If you do not find reviewers before May, ping me again in May which is usually a calmer period. Thanks again and sorry for not being able to review the submission now.

**Comment by oliviaguest** [comment]
@amyzhang and @xuedong would either/both of you be interested and free to review this? ☺️

**Comment by oliviaguest** [comment]
@raar1 do you think you might be able to review this? 😊

**Comment by oliviaguest** [comment]
@D-Maurel I apologise this is taking so long — my impression here is that most people are not checking their notifications, but I am emailing them or at least trying to locate their email addresses too. No luck so far, sadly. I am hoping things will improve. 🤞

**Comment by Kahss** [comment]
@oliviaguest Thank you for taking care of it, I'm not in a rush :). I hope things will improve !

**Comment by oliviaguest** [comment]
Thank you, @wmkouw, for being a reviewer. 😊

**Comment by oliviaguest** [comment]
@D-Maurel do you have any other ideas for who might be a good person to ask? If yes, please feel free to give me their names (ideally without tagging them). ☺️

**Comment by rougier** [comment]
Any progress?

**Comment by oliviaguest** [comment]
I am trying my best  — as this is not really my primary (or even secondary) area of research, I am trying all the avenues I can. If anybody has recommendations, that would be appreciated. Pandemic and things relating to it, seem to be making it harder for people to have the time; which is utterly understandable.

**Comment by oliviaguest** [comment]
Ah, also... FWIW @wmkouw has agreed, but I think I cannot "assign" them without them commenting on this issue?

**Comment by wmkouw** [comment]
Hi, sorry for not responding. I did promise Olivia that I would review this paper. I have not been very communicative because I just came back from birth leave and I'm swamped with work.

I told Olivia that I could make time in week 22 (May 31th until June 4th). Hope that's okay.

**Comment by Kahss** [comment]
Thank you @oliviaguest  and @wmkouw for dealing with my article :slightly_smiling_face: !

**Comment by oliviaguest** [comment]
I think it would be ideal to have one more reviewer — and yes, 100% OK with me for you (@wmkouw) to take a reasonable amount of time given your constraints. 😊

**Comment by wmkouw** [comment]
Hi, a small update: I read the paper on Wednesday and briefly went through the paper it is reproducing. I just need to go through the code to complete my review, but I haven't been able to run the examples yet. When I generate local virtualenv directories and run pip, the dependencies are being installed in my conda env directory instead of the local ones. I need to figure out how to fix that, but I've had back-to-back meetings since Wednesday.

TL;DR My review isn't done yet; I will go through the code next week. Hope that's okay.

**Comment by oliviaguest** [comment]
> When I generate local virtualenv directories and run pip, the dependencies are being installed in my conda env directory instead of the local ones. I need to figure out how to fix that, but I've had back-to-back meetings since Wednesday.

@D-Maurel can you please help @wmkouw (see my quote in this comment) to use your code, as this is central to reproducibility. Make the changes needed to code and documentation to help others too, etc., of course. Thanks! ☺️

**Comment by wmkouw** [comment]
Update: we had a family medical emergency and I've been at the hospital last week. I am back to work but I am swamped with teaching this week. I will finish the review next week.

I'm sorry. Things don't always go according to plan.

**Comment by Kahss** [comment]
@wmkouw Thank you for taking the time to read my article. I'm really sorry to hear about your medical emergency, I hope everything went fine. I will of course help you to solve any problem related to the execution of my code, please ask any question you have :) .

**Comment by oliviaguest** [comment]
@wmkouw I am so sorry to hear this — and, please, take your time. 🌸

**Comment by oliviaguest** [comment]
@ESSYCHAN has agreed to review this as well. ☺️ 

Esther, if you can drop a comment below, so I can assign you that would be great — thanks.

**Comment by wmkouw** [comment]
I've managed to find some time to finish my review. Instead of integrating conda en virtualenv, I just grabbed a Docker container and worked from there. Everything runs as advertised.

---

Summary:
This papers successfully replicates an influential technique for program debloating. The original implementation was written in C, whereas this one is in Python. It includes tests for the debloater on Python code and reports additional quantitative results.

Strengths:
- A python version of the debloater is valuable and demonstrates the language-agnostic aspect of CHISEL.
- The paper discusses several key practical considerations that were not reported in the original paper, for example the requirement of multiple runs (Sec 2.3).
- The paper explores the use of existing unit tests instead of a test script that would have to be written prior to debloating.
- Table 2 adds an important perspective: the original paper reports large reductions in the number of statements (Figure 8) for CHISEL. However, this Table shows a less than 10% gain in process duration over no RL, which is much more modest.
- The paper is well written.

Weaknesses:
- Relying on UT can be tricky in practice, I think. In Section 2.2, it says: "However, this approach requires the software test suite to cover all the required functionalities." Suppose a co-worker leaves me with their large (and probably bloated) code base without telling me whether the tests have sufficient coverage. Should I just run the debloater? Or should I inspect the code thoroughly first, in case the debloater accidentally removes important functionality?

Minor points:
- (Sec. 2.3, RL) Perhaps worth adding: for gamma = 0, the RL agent defaults to a multi-armed bandit (performing 1 action and immediately observing reward; https://en.wikipedia.org/wiki/Multi-armed_bandit).
- (Sec. 3.2) ".. a while test suite". A "wide test suite"?
- (Conclusion) ".. outperform industrial debloating tools." I assume this refers to C-REDUCE and PERSES?

---

I apologize for my somewhat shallow review; I should have probed the code more thoroughly (I only ran the examples). It's been a trying time and this is all the effort I can spare at the moment.

**Comment by ESSYCHAN** [comment]
Hello @oliviaguest, 
Yes I have agreed that I'll review this paper. 

Should I state the timeline on when I expect to finish?

**Comment by oliviaguest** [comment]
@ESSYCHAN that would be very useful and kind of you. 😊

**Comment by oliviaguest** [comment]
@wmkouw we appreciate your time and are very thankful. I hope things are as well as they can be. 🌼

**Comment by ESSYCHAN** [comment]
I'm committing to give a proper review by 19th July.

**Comment by Kahss** [comment]
Hello, I haven't noticed the notification mentionning that a review was available. I hope the delay hasn't done any harm, if not I'm really sorry.

Thank you @wmkouw for your review. Your point about the UT related weakness is totally relevant, and I will modify the original paper according to it, introducing further more nuances about the use of UT.

Thank you @ESSYCHAN for accepting to review my paper. I'm looking forward to read your comments.

**Comment by Kahss** [comment]
Hello all. I would like to know if you have any news ?

**Comment by oliviaguest** [comment]
@ESSYCHAN @wmkouw are your reviews possible to be done by the end of the month?

**Comment by rougier** [comment]
@oliviaguest Any update ?

**Comment by oliviaguest** [comment]
Nothing from me sadly... I tried 2 weeks ago as you see above. I will now try to email the reviwers.

**Comment by oliviaguest** [comment]
@D-Maurel have you addressed @wmkouw's comments and/or do you need feedback from them at the moment?

**Comment by Kahss** [comment]
@oliviaguest I haven't, I was waiting for another possible review to address every feedback at once.

**Comment by Kahss** [comment]
Hi all, @oliviaguest, would you have any update for a possible second review ?

**Comment by oliviaguest** [comment]
I'm really struggling to get anything from the other reviewer. Do you have
any suggestions for a new 2nd reviewer? That might be the only way forward.

**Comment by Kahss** [comment]
I don't sorry :/ . What is the minimum number of reviewer required ?

**Comment by ESSYCHAN** [comment]
Hello Denis,

I'm currently working on the review. should be able to give you feedback by
close of business today.

Best wishes,
Esther.

**Comment by Kahss** [comment]
Hello Esther,

Thank you very much ! I'm looking forward to reading your comments !

**Comment by ESSYCHAN** [comment]
Dear Maurel,

Attached, please find the PDF document with my comments on the paper:



[Replication_CHISEL (1).pdf](https://github.com/ReScience/submissions/files/7476387/Replication_CHISEL.1.pdf)



I'll also issue a few comments here in this section:

I generally enjoyed reviewing your work, I think it is a commendable piece of work but there is room to add a few things.

**The paper**

It was an easy and straightforward read that gave me an overview about what has been done previously and what you currently are working on:
A few things to note is 

> Clearly state your problem and how different it is from the previous implementation, I felt like there is more to do here and justify your work.
> Support most of the claims by expounding on your references.
> State as to why using Python as opposed to C (as previously done) would be beneficial for an RL task. This did not come out clearly (see comments in the paper)
>It is important to clarify about the objectives that you were working on, otherwise, it was hard to distinguish this from the replication.
>Is there a way to prove the quality of the UT cases (you will see comments in the pdf document)
>For the results, I suggested a few graphs for the reader to see the gain and improvement
>For the conclusion, give areas to work on in the future and a good abstract summarizing succinctly and expressively What the paper is about
>This comment was left out on the  pdf about the first paragrph : (The introduction section needs to have several references. When you mention factors, it is important always to ascertain as to why such claims are being made)


**Code**

> The code was a straightforward run other than struggling with setting up the Venv environment: The readme document should include details for different scenarios of V.env, this will help users not run into setup issues
> The test cases set up were easy to run given the commands, they all successfully passed s indicated by the results in the paper.
> I encountered a few errors while running the Optimus __main__file() on its own, although the test ran fine and successfully. 
File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/genericpath.py", line 42, in isdir
    st = os.stat(s)
TypeError: stat: path should be string, bytes, os.PathLike or integer, not NoneType
>I also noted that the test cases were resetting to range (3) on their own even after changing them to a different generator expression - You may need to check on this.
> the command $make example seems not to work properly
Error received make: *** No rule to make target `example'.  Stop.
>I particularly commend you on the code organization it was very clear to follow

With these few remarks and those on the pdf document, I wish you the best. 

Let me know in case you need any clarification on my part.

Best wishes,
Esther

**Comment by Kahss** [comment]
Thank you very much Esther, your comments are really insightful. I will work on the paper in the few next days and come back with an updated version. Thanks again !

**Comment by Kahss** [comment]
Hello again @ESSYCHAN ,
Thanks again for your comments. I've been integrating them as precisely as possible in the paper sources. The gitlab repository has been adequately updated.

I would have some questions about your remarks on the code:
- When you say that the __main__.py file produces an error, could you please give me the command you're trying to run ? The file is not meant to be run directly, but rather through the indirect command $make run_example TARGET='<target>' which itself calls $python -m optimus -c examples/<target>/config.yaml
- I'm not sure to understand your remarks regarding the test cases resetting to range (3). Could you please detail it ?
- About the command $make example, it's possibly a typo, the right command being $make example**s**. Could you please confirm that it's just a typo problem ?
- Thank you very much for your kind return. Lots of efforts have been put in the code organization.

**Comment by Kahss** [comment]
Dear @wmkouw , your comments have also been integrated, thanks again.

**Comment by wmkouw** [comment]
> @ESSYCHAN @wmkouw are your reviews possible to be done by the end of the month?

I'm sorry. I don't have the time to review the paper again. I recommended to accept the paper the first time around and I trust @ESSYCHAN's feedback to have led to improvements, so I'd recommend 'accept' again.

**Comment by ESSYCHAN** [comment]
> Hello again @ESSYCHAN , Thanks again for your comments. I've been integrating them as precisely as possible in the paper sources. The gitlab repository has been adequately updated.
> 
> I would have some questions about your remarks on the code:
> 
> * When you say that the **main**.py file produces an error, could you please give me the command you're trying to run ? The file is not meant to be run directly, but rather through the indirect command $make run_example TARGET='' which itself calls $python -m optimus -c examples//config.yaml
> * I'm not sure to understand your remarks regarding the test cases resetting to range (3). Could you please detail it ?
> * About the command $make example, it's possibly a typo, the right command being $make example**s**. Could you please confirm that it's just a typo problem ?
> * Thank you very much for your kind return. Lots of efforts have been put in the code organization.

Hello,

For the main file, I tried just running the file on its own, but it worked well when I ran it indirectly. $make run_example TARGET
yes this was a typo problem, it's running well with $make example**s**
As for the range(3), I tried to increase to range 10 but as soon as the file runs, the range resets to the original value 3.

I believe that given the corrections you've made I would recommend a pass again.

**Comment by Kahss** [comment]
@wmkouw Thank you very much for your approval.

@ESSYCHAN Ok I understand. Actually to be sure that examples can be run whatever the potential previous execution errors, I use a `<filename>_CLEAN.py` version for the ressource files (in the `ressource` folder). This way, to ensure that everything is ok, I erase the last copy of the file and I duplicate the `CLEAN` version. I hope it helps.
And anyway, thank you very much for your approval.

**Comment by Kahss** [comment]
@oliviaguest , could you please tell me what is the next step of the publication process ? Do you require a third reviewer for the paper to be accepted ?

**Comment by rougier** [comment]
@D-Maurel We need only two reviewer such that you paper is virtually accepted but @oliviaguest will confirm (or not)

**Comment by Kahss** [comment]
Thanks for the precisions @rougier. I can't wait to see the acceptance confirmed !

**Comment by rougier** [comment]
@oliviaguest Can we proceed with publication?

**Comment by oliviaguest** [comment]
@rougier I have been and still am very unwell for the past 3 weeks. I can try and get to this soon, but unlikely to be today as I am playing catch up. Apologies.

**Comment by rougier** [comment]
@oliviaguest No problem. As soon as we got your go, I can publish the article.

**Comment by oliviaguest** [comment]
> @wmkouw Thank you very much for your approval.
> 
> @ESSYCHAN Ok I understand. Actually to be sure that examples can be run whatever the potential previous execution errors, I use a `<filename>_CLEAN.py` version for the ressource files (in the `ressource` folder). This way, to ensure that everything is ok, I erase the last copy of the file and I duplicate the `CLEAN` version. I hope it helps. And anyway, thank you very much for your approval.

Is this clear now in the documentation, @D-Maurel?

**Comment by Kahss** [comment]
@oliviaguest , yes it is now. I should have modified the README earlier, my bad.

**Comment by oliviaguest** [comment]
@D-Maurel @ESSYCHAN is there anything left unaddressed from Esther's feedback here? If not, @ESSYCHAN could you check things that have been made sense for you? Your feedback has been super valuable and I'd hate for it not be used to its full extent. ☺️

**Comment by Kahss** [comment]
@oliviaguest , for me everything has been addressed. @ESSYCHAN please tell me if there still is any unclear spot in the paper or in the code.

**Comment by rougier** [comment]
@oliviaguest If we do not get news, you can make a decision.

**Comment by ESSYCHAN** [comment]
My apologies Team, I'll run the experiments once more and give feedback
this evening or latest tomorrow morning. I understand everyone has been
busy trying to close the year, same for me here. It is hectic but pusing to
close off the review as well.

Best regards,
Esther.

**Comment by rougier** [comment]
@oliviaguest @ESSYCHAN Happy new year everyone. Any update on this?

**Comment by ESSYCHAN** [comment]
Hello and a happy new year to you!!

I have run in from my end, everything seems to be working fine on my end.

Additional comments (just one :) )
maybe add **  brew install cmake .. in the installation guidelines for
those that don't have cmake


best wishes,
Esther.

**Comment by Kahss** [comment]
Hello and happy new year everyone :) !

Glad to hear that everything works well, thank you again @ESSYCHAN for your comments and feedback.
The `README` has been updated with a pointer to a Stack Overflow post explaining how to run a `Makefile` in case you are on a Windows machine.

Best wishes,
Denis Maurel

**Comment by rougier** [comment]
Cool. If @oliviaguest is ok, we can proceed with publication.

**Comment by Kahss** [comment]
Hi @rougier , @oliviaguest . Hope you're doing great. Do you have any update ?

**Comment by oliviaguest** [comment]
@D-Maurel Apologies for being so late about this, I keep getting unfortunate medical things happening. I think this is fine to process with publication. Has everything been filled in the metadata files (see: https://github.com/rescience/articles#readme) for the process to start? 😊

**Comment by Kahss** [comment]
I've updated the metadata file, and I've started to follow the process, but after having successfully reserved a DOI on the sandbox Zenodo server, I have a problem running the same command with the `--zenodo` flag :
```
> ./process.py --zenodo --metadata metadata.yaml --pdf article.pdf 
Traceback (most recent call last):                                                                                                                                                                                 
  File "./process.py", line 77, in <module>                                                                                                                                                                        
    article_id, article_doi = reserve_doi(server, token)                                                                                                                                                           
  File "./process.py", line 21, in reserve_doi                                                                                                                                                                     
    response.json()["message"])                                                                                                                                                                                    
  File "/home/xxx/.virtualenvs/chisel-replication/lib/python3.8/site-packages/requests/models.py", line 910, in json                                                                                 
    return complexjson.loads(self.text, **kwargs)                                                                                                                                                                  
  File "/usr/lib/python3.8/json/__init__.py", line 357, in loads                                                                                                                                                   
    return _default_decoder.decode(s)                                                                                                                                                                              
  File "/usr/lib/python3.8/json/decoder.py", line 337, in decode                                                                                                                                                   
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())                                                                                                                                                              
  File "/usr/lib/python3.8/json/decoder.py", line 355, in raw_decode                                                                                                                                               
    raise JSONDecodeError("Expecting value", s, err.value) from None                                                                                                                                               
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0) 
```
Do you have any recommendation to continue ?

**Comment by Kahss** [comment]
Nevermind, after checking the request with Insomnia, it was just a problem of a account verification.

**Comment by Kahss** [comment]
My bad, as @ESSYCHAN has provided the link to the publication README, I thought I had to do it myself. I just stopped everything, and removed my post from the paper number claim issue. Everything is up to date in the metadata file :).

**Comment by Kahss** [comment]
Anyway, I hope you get better soon @ESSYCHAN !

**Comment by oliviaguest** [comment]
Oh, oops. I didn't mean for you to do the publication yourself. Just to check the metadata. 😅 

I will get this published for you @D-Maurel, as soon as I can. Thanks for helping me to get it started.

**Comment by oliviaguest** [comment]
@D-Maurel I noticed you haven't updated the metadata in a while (11 months) or compiled the PDF? Sorry, I am rusty with this procedure, but I think you have to compile it for me? Also... @ReScience/editors are there any special requirements if this is on GitLab?

**Comment by Kahss** [comment]
Hi @oliviaguest ,
The PDF is two month old on the gitlab repository, could you please confirm me that this is what you are seeing ?

![Capture d’écran de 2022-01-31 14-28-18](https://user-images.githubusercontent.com/4797989/151802006-8f897644-ceb9-4e42-91e2-4889cd3b4bdc.png)

For the metadata, I thought I had filled everything that needed to. The part that is still empty is said to be filled by the editor. I assumed it was you but it was maybe a mistake ? Everything is complete except this part

```
# This information will be provided by the editor
dates:
  - received:  November 1, 2018
  - accepted:
  - published: 

# This information will be provided by the editor
article:
  - number: # Article number will be automatically assigned during publication
  - doi:    # DOI from Zenodo
  - url:    # Final PDF URL (Zenodo or rescience website?)

# This information will be provided by the editor
journal:
  - name:   "ReScience C"
  - issn:   2430-3658
  - volume: 4
  - issue:  1
```

**Comment by rougier** [comment]
You'll need to change the volume/issue as well. But the editor is mostly responsible for doing. @oliviaguest tell me if you need help on the publication.

**Comment by oliviaguest** [comment]
@rougier @Kahss I am so sorry to be taking so long. I will try to get this done today.

**Comment by Kahss** [comment]
@oliviaguest Absolutely no problem, I really hope that you're seeing the end of your medical things. Take care.

**Comment by oliviaguest** [comment]
@ESSYCHAN what is your orcid? @wmkouw I have found yours: https://orcid.org/0000-0002-0547-4817 😊

**Comment by ESSYCHAN** [comment]
Hi Olivia,

 https://orcid.org/0000-0002-1327-1198

**Comment by oliviaguest** [comment]
@ESSYCHAN thank you!

**Comment by ESSYCHAN** [comment]
Thanks Olivia. I hope you are feeling much better xx :)

**Comment by oliviaguest** [comment]
@Kahss I think I now remember how to do this. 😅  Can you either allow me (same username as GitHub) to push to your repo on GitLab or update the `metadata.yaml` using this, please? And check everything is correct and then compile a new PDF for your paper called `article.pdf`?

```
# To be filled by the author(s) at the time of submission
# -------------------------------------------------------

# Title of the article:
#  - For a successful replication, it shoudl be prefixed with "[Re]"
#  - For a failed replication, it should be prefixed with "[¬Re]"
#  - For other article types, no instruction (but please, not too long)
title: "[Re] Effective Program Debloating via Reinforcement Learning"

# List of authors with name, orcid number, email and affiliation
# Affiliation "*" means contact author
authors:
  - name: Denis Maurel
    orcid: 0000-0001-5658-5923
    email: d.maurel@groupeonepoint.com
    affiliations: 1,*

  - name: Jérôme Fillioux
    orcid: 0000-0001-7304-1899
    email: j.fillioux@groupeonepoint.com
    affiliations: 1

  - name: Dan Gugenheim
    orcid: 0000-0003-4800-7909
    email: d.gugenheim@groupeonepoint.com
    affiliations: 1

# List of affiliations with code (corresponding to author affiliations), name
# and address. You can also use these affiliations to add text such as "Equal
# contributions" as name (with no address).
affiliations:
  - code:    1
    name:    onepoint sud-ouest R&D
    address: Bordeaux, France

# List of keywords (adding the programming language might be a good idea)
keywords: program debloating, reinforcement learning, unit test, python, machine learning, decision tree, delta debugging

# Code URL and DOI (url is mandatory for replication, doi after acceptance)
# You can get a DOI for your code from Zenodo,
#   see https://guides.github.com/activities/citable-code/
code:
  - url: https://gitlab.com/onepoint/recherche-et-developpement/optimus/chisel-replication
  - doi: 10.5281/zenodo.4530318

# Date URL and DOI (optional if no data)
data:
  - url:
  - doi:

# Information about the original article that has been replicated
replication:
 - cite: Heo, K., Lee, W., Pashakhanloo, P., & Naik, M. (2018, October).Effective program debloating via reinforcement learning. In Proceedings of the 2018 ACM SIGSAC Conference on Computer and Communications Security (pp. 380-394).
 - bib:  heo2018effective
 - url:  https://www.cis.upenn.edu/~mhnaik/papers/chisel.pdf
 - doi:  10.1145/3243734.3243838


# Don't forget to surround abstract with double quotes
abstract: "This paper presents a replication of the CHISEL debloating system with some clarifications on caveats of the original method.
           It also presents a new property testing method using unit tests.
           The original results are extended with numerical proves of the improvements linked to the Reinforcement
           Learning part of the CHISEL system compared to the original HDD system.
           Several Python librairies are debloated and the improvements in terms of number of iterations and time consumption are presented.
           The results show that the RL part of CHISEL significantly improves the original HDD algorithm."

# Bibliography file (yours)
bibliography: bibliography.bib

# Type of the article
# Type can be:
#  * Editorial
#  * Letter
#  * Replication
type: Replication

# Scientific domain of the article (e.g. Computational Neuroscience)
#  (one domain only & try to be not overly specific)
domain: Program Debloating

# Coding language (main one only if several)
language: Python


# To be filled by the author(s) after acceptance
# -----------------------------------------------------------------------------

# For example, the URL of the GitHub issue where review actually occured
review:
  - url: https://github.com/ReScience/submissions/issues/48

contributors:
  - name: Olivia Guest
    orcid: 0000-0002-1891-0972
    role: editor
  - name: Wouter Kouw
    orcid: 0000-0002-0547-4817
    role: reviewer
  - name: Esther Mulwa
    orcid: 0000-0002-1327-1198
    role: reviewer

# This information will be provided by the editor
dates:
  - received: February 10, 2021
  - accepted: January 18, 2022
  - published: February 24, 2022

# This information will be provided by the editor
article:
  - number: 12 # Article number will be automatically assigned during publication
  - doi: 10.5281/zenodo.6255131 # DOI from Zenodo
  - url: https://zenodo.org/record/6255131/files/article.pdf # Final PDF URL (Zenodo or rescience website?)

# This information will be provided by the editor
journal:
  - name:   "ReScience C"
  - issn:   2430-3658
  - volume: 7
  - issue:  1
```

**Comment by oliviaguest** [comment]
@rougier I *think* I am doing this right, but please ignore/jump in as needed. 😆 Thanks. And sorry for being so rusty.

**Comment by Kahss** [comment]
@oliviaguest @rougier I've just updated the metadata.yaml file, however when I try to run the `make` command as mentioned in the README, the compilation ends with this error : 

```
! LaTeX Error: File `biblatex.sty' not found.                                                                                                                                                                      
                                                                                                                                                                                                                   
Type X to quit or <RETURN> to proceed,                                                                                                                                                                             
or enter new name. (Default extension: sty)                                                                                                                                                                        
                                                                                                                                                                                                                   
Enter file name:                                                                                                                                                                                                   
! Emergency stop.                                                                                                                                                                                                  
<read *>                                                                                                                                                                                                           
                                                                                                                                                                                                                   
l.308 \RequirePackage                                                                                                                                                                                              
                     {bibentry}^^M                                                                                                                                                                                 
No pages of output.                                                                                                                                                                                                
Transcript written on article.log.                                                                                                                                                                                 
Latexmk: Missing input file: 'biblatex.sty' from line                                                                                                                                                              
  '! LaTeX Error: File `biblatex.sty' not found.'                                                                                                                                                                  
Latexmk: 'pdflatex': source file 'biblatex.sty' doesn't exist. I'll try making it...                                                                                                                               
------------                                                                                                                                                                                                       
Running 'make "biblatex.sty"'                                                                                                                                                                                      
------------                                                                                                                                                                                                       
make[1] : on entre dans le répertoire « /home/d.maurel/.virtualenvs/chisel-replication »                                                                                                                  
make[1]: *** Aucune règle pour fabriquer la cible « biblatex.sty ». Arrêt.                                                                                                                                         
make[1] : on quitte le répertoire « /home/d.maurel/.virtualenvs/chisel-replication »                                                                                                                      
Latexmk: Errors, so I did not complete making targets                                                                                                                                                              
Collected error summary (may duplicate other messages):                                                                                                                                                            
  pdflatex: Command for 'pdflatex' gave return code 256                                                                                                                                                            
Latexmk: Use the -f option to force complete processing,                                                                                                                                                           
 unless error was exceeding maximum runs of latex/pdflatex.                                                                                                                                                        
Makefile:30: recipe for target 'article.pdf' failed                                                                                                                                                                
make: *** [article.pdf] Error 12
```

Maybe I'm missing something ? So far I've always used the overleaf compilation tool to get a pdf result, so I hadn't those problems.

**Comment by rougier** [comment]
You can try to type `biber article` on the command line.

**Comment by Kahss** [comment]
I've installed `biber`, run it and run `make` again, now I'm getting an `article.pdf` output, however I think there is still some problems as the received date is still "XX December 2020" and I get this output :

```
Output written on article.pdf (6 pages).
Transcript written on article.log.
Latexmk: Found input bbl file 'article.bbl'
Latexmk: Log file says output to 'article.pdf'
Latexmk: Found biber source file(s) [article.bcf bibliography.bib]
Latexmk: Errors, so I did not complete making targets
Collected error summary (may duplicate other messages):
  pdflatex: Command for 'pdflatex' gave return code 256
Latexmk: Use the -f option to force complete processing,
 unless error was exceeding maximum runs of latex/pdflatex.
Makefile:30: recipe for target 'article.pdf' failed
make: *** [article.pdf] Error 12
```

@rougier Any idea ?

**Comment by rougier** [comment]
Try a `make clean`, `make`, `biber article` and `make` again. If this does not work, you can try manual compilation.

**Comment by Kahss** [comment]
@rougier Ok so after investigation, there was a problem with the `metadata.tex` which was obsolete. Deleting it and running again `make metadata.tex` fixed the problem. There is a final version of `article.pdf` in the repository.

I was afraid there was an error regarding the insertion of a bib key just after "A partial replication of" in the article header, but when I compared to the latest published article, it seems to be the desired behavior. I didn't know if it was intended, so I wanted to mention it, just in case.

**Comment by rougier** [comment]
Well no, it is not the intended behavior. The bib key should be replaced with the actual citation. Which article did you check?

**Comment by Kahss** [comment]
Those ones:
- https://zenodo.org/record/5718075/files/article.pdf
- https://zenodo.org/record/5657320/files/article.pdf

**Comment by rougier** [comment]
AAAAaaaaahhhhhhh I'm among the co-authors, what a shame! So it's not supposed to be like this, I need to correct it. In your case, probably the simpler is to insert the full reference or to fix the script. But the citation should be processed by latex. Not sure why it is not.

**Comment by Kahss** [comment]
@rougier I've updated my pdf, everything should be ok. To solve the problem, I just had to change `\replicationBIB` line 94 and 95 in `header.tex` by `\fullcite{\replicationBIB}`. Hope it might help.

**Comment by rougier** [comment]
Cool. @oliviaguest I think the paper might be ready for publication.

**Comment by oliviaguest** [comment]
I am confused. Is the pdf final?

**Comment by Kahss** [comment]
Hello @oliviaguest. Yes it is, it has been updated with all the metadata :) .

**Comment by oliviaguest** [comment]
@Kahss I will try again soon, sorry. I always struggle with this, perhaps because I am on a Mac like a loser. 😆

**Comment by Kahss** [comment]
@oliviaguest Thanks a lot, I'm it's almost the end of the tunnel :smile:  !

**Comment by oliviaguest** [comment]
@Kahss oh yeah, here we go, haha. Can you please do me a massive favour and pick a date in the future for "published" like next Monday and I will do it then. Because otherwise if I publish it now it will have a date in the past. So sorry. ☺️

**Comment by Kahss** [comment]
@oliviaguest The metadata and `article.pdf` have been updated :).

**Comment by Kahss** [comment]
Hi @oliviaguest , I hope you're doing well. Do you have any update about the publication :) ?

**Comment by oliviaguest** [comment]
Sorry for not being clear. Can you compile the pdf with metadata for published date to be next Monday, or so? So I can publish it with the right date, please?

**Comment by Kahss** [comment]
@oliviaguest Maybe it was I who was not clear, 6 days ago I changed the publication date in the metadata file and recompile the pdf with the new value.
![Capture d’écran de 2022-03-23 09-53-31](https://user-images.githubusercontent.com/4797989/159660692-3fa04328-6c1b-4a7b-9206-14521db6b4ae.png)
I have changed it again, and now the publication date is set to March 28 2022. Please tell me if anything's still unclear :) .

**Comment by Kahss** [comment]
@oliviaguest Maybe it was I who was unclear. 6 days ago, I changed the value in the metadata file, and recompile the pdf with the new value. 
I have changed it again, and now the publication date is set to March 28. Please tell me if there's still something unclear :) .

**Comment by oliviaguest** [comment]
@Kahss I am having so much trouble — BUT! — I think I did it. Can somebody help me out to check I hacked my way to actually managing it? @rougier @khinsen 😊

**Comment by rougier** [comment]
Yes, what is the problem exactly?

**Comment by Kahss** [comment]
Hello @oliviaguest, any news ?

**Comment by oliviaguest** [comment]
@rougier is it properly published? ☺️

**Comment by jfillioux** [comment]
Hi,

Reading the publish process here :
https://github.com/ReScience/articles

I understand that the last step for publishing a new article is :
"14. Finally, copy the contents of article.bib from the doi folder for this paper into [rescience.github.io/_bibliography/published.bib] **and send a final pull request** (You can do this from the web)."

I can see here that the copy in the artcile.bib has been done : 
https://github.com/ReScience/rescience.github.io/blob/sources/_bibliography/published.bib

**but then I can't find the associated pull request here :**
https://github.com/ReScience/rescience.github.io/pulls?q=is%3Apr

Maybe that's the last missing step ?

Thx for your help

**Comment by oliviaguest** [comment]
@jfillioux yeah, I don't know, sorry. I hope @khinsen or @rougier can check for us.

**Comment by rougier** [comment]
My bad, I did not updated the website. It's now done but then I realized the information for this article should be volume 8, issue 1, number 1. I modified the bibtex on the website and @oliviaguest can modify the Zenodo entry (using the online interface). This should not modify the DOI.

**Comment by Kahss** [comment]
Thank you very much for everything, glad to see the article online :smile: !
